drivers/usb/udc/udc_stm32.c

```diff
@@ -41,26 +41,42 @@ LOG_MODULE_REGISTER(udc_stm32, CONFIG_UDC_DRIVER_LOG_LEVEL);
 #define UDC_STM32_IRQ		DT_INST_IRQ_BY_NAME(0, UDC_STM32_IRQ_NAME, irq)
 #define UDC_STM32_IRQ_PRI	DT_INST_IRQ_BY_NAME(0, UDC_STM32_IRQ_NAME, priority)
 
-#define USB_OTG_HS_EMB_PHY (DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usbphyc) && \
+#define USB_OTG_HS_EMB_PHYC (DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usbphyc) && \
+			     DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
+
+#define USB_OTG_HS_EMB_PHY (DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_otghs_phy) && \
 			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
 
 #define USB_OTG_HS_ULPI_PHY (DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy) && \
 			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
 
 /**
- * The following defines are used to map the value of the "maxiumum-speed"
- * DT property to the corresponding definition used by the STM32 HAL.
+ * Select the desired speed based on enabled high speed support and phy.
+ *
+ * 1.  high speed support enabled and hs phy is used
+ * 1.1 maximum speed is reduced by devicetree configuration
+ * 1.2 maximum speed is high speed
+ * 2.  high speed phy is used, but high speed support is not enabled
+ * 3.  non-hs capable hardware, but with st_stm32_usb compatible
+ * 4.  everything else
  */
-#if defined(CONFIG_SOC_SERIES_STM32H7X) || USB_OTG_HS_EMB_PHY
-#define UDC_STM32_HIGH_SPEED             USB_OTG_SPEED_HIGH_IN_FULL
+#if CONFIG_UDC_DRIVER_HIGH_SPEED_SUPPORT_ENABLED && \
+	(USB_OTG_HS_EMB_PHYC || USB_OTG_HS_EMB_PHY || USB_OTG_HS_ULPI_PHY)
+/* as UDC_BUS_SPEED_HS is an enum we can not directly use is in the following
+ * preprocessor logic. make sure it has the expected value during compilation.
+ */
+BUILD_ASSERT(UDC_BUS_SPEED_HS == 2, "UDC_BUS_SPEED_HS is expected to be 2");
```

Comment:
I think the BUILD_ASSERT() is not necessary here, UDC_BUS_SPEED_HS has nothing to do with DT maximum_speed property.